Added constexpr specific code for CUDA compatibility

vflorins opened the pull request:
The changes should not affect anything when not using CUDA, but just in case please check whether the tests run correctly. Lucius, see if you agree with the constexpr features.

vflorins committed:
Added constexpr specific code for CUDA compatibility

jgaloguz committed:
Update background_server_batl.cc

Fixed wrong name of parent class in copy constructor.

jgaloguz reviewed (approved):
The benchmark tests all run correctly. Note that I piggy-backed on this pull request to fix an unrelated bug in BackgroundServerBATL that we've known about but haven't fixed for a while.

jgaloguz committed:
Major update to benchmarks

The benchmarks codes have been updated to be more inform and the script for automated testing has been improved.
 - An optional argument to run_tests.sh allows for testing with less than the total number of CPUs available. The default number of CPUs is still the maximum.
 - The automated testing now generates log files which separate each benchmark into configuration, compilation, execution, and plotting stages, reporting the output and possible errors within each.
 - Plots are generated for each benchmark to quickly assess the correct execution of the code.

jgaloguz commented:
Actually, there was one test that failed silently because the "semi-automated" testing script I wrote had some deficiencies. I fixed this but it prompted me to improve the benchmarks folder (as a whole). Please refer to the last commit in this branch to see specifics. This brings the `run_tests.sh` script to be one step below fully-automated testing, since it only requires a visual check for each of the plots. From this point, it should be easy to construct scripts that check the correctness of the output without needing a visual inspection through error means, variances, fits, etc.

LuciusSchoenbaum reviewed (approved):
The updates to fundamental classes for CUDA and constexpr const-ness look ok to me. 

Two minor comments: 
- **common/simple_array.hh:34** `is_simple_array` could be implemented differently in C++20 using `std::derived_from<Derived, Base>`, some might argue this is a cleaner solution. I didn't look carefully so take it as a suggestion. 
- **common/gpu_config.hh:41** typo in the print string, should be "performance".

vflorins commented:
The print statement containing the word "performance" was removed in this commit because the `cudaDeviceProp` structure no longer has that data member.
